Adding performance functional tests

dorzheh opened the pull request:
# Description

Adding more performance tests to the E2E performance test suite

Fixes # (issue)

## Type of change

Please select the appropriate options:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] cd <repo_name>/features and execute FEATURES=performance make functests

**Test Configuration**:

- Versions: 4.3.0-0.nightly-2019-12-13-180405
- Hardware: baremetal

## Checklist

- [x] I have performed a self-review of my own code
- [ ] If a change is adding a feature, it should require a change to the README.md and the review should catch this.
- [ ] If the change is a fix, it should have an issue. The review should make sure the comments state the issue (not just the number) and it should use the keywords that will close the issue on merge.
- [ ] A change should not being merged unless it passes CI or there is a comment/update saying what testing was passed.
- [ ] PRs should not be merged unless positively reviewed.

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -12,36 +13,47 @@ import (
+const (
+	perfTestNamespace            = "perfomancetest"
+	perfWorkerNodesLabel         = "node-role.kubernetes.io/worker-rt="
+	perfWorkerRtKernelMcTemplate = "11-worker-rt-kernel"
```
youll  need to align to the changes introduced in  https://github.com/openshift-kni/baremetal-deploy/pull/115

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +62,160 @@ var _ = Describe("performance", func() {
+	})
+
+	var _ = Context("Network latency parameters", func() {
+		It("Should contain configuration injected through the 00-tuned-network-latency template", func() {
```
We are dependent on the tuned-network-latency profile here, for this repo it might be ok.
@cynepco3hahue  any thoughts here ?

cynepco3hahue replied:
Our worker-rt tuned profile depends on it, so I think it should be ok

yanirq commented:
/ok-to-test

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -78,7 +230,7 @@ func mcdForNode(node *k8sv1.Node) (*k8sv1.Pod, error) {
 		return nil, err
 	}
 	// there should be one machine config deamon per node
```
remove the comment

yanirq replied:
remove the comment here if its in the failure description

yanirq commented:
@dorzheh  I would like to see the prow job running on this PR for performance tests before merging it, Can you edit one of the files under feature/performance/*  with a change  just to see the tests run?  and then revert that change ?

dorzheh commented:
> @dorzheh I would like to see the prow job running on this PR for performance tests before merging it, Can you edit one of the files under feature/performance/* with a change just to see the tests run? and then revert that change ? (once we see the e2e tests pass)

@yanirq , pretty weird to me that this is the only way to trigger the job but added a dummy file that will be removed later on.

yanirq commented:
> @yanirq , pretty weird to me that this is the only way to trigger the job but added a dummy file that will be removed later on.

This should be fixed in the ci job (openshift/release)  ill update it for future PRs

dorzheh commented:
@yanirq ,the previous commit solved the following issue: it seems 4.3 uses tuned 2.11 and it has an issue with the latency parsing (for example _set_latency expects an integer but receives string that cannot be "casted" to int).  According to the tuned code its fixed in v2.13+ . Hence i commented force_latency in the template. Afterwards, the tests have passed except the one checking "rt" suffix in the kernel name. Now i see the tests , that previously passed successfully , failed again. For me it  looks like the following happened: when tuning templates applied , tuned polling the changes within appropriate amount of time. It seems that tests was running before the changes were injected by the tune daemon.

yanirq commented:
/retest

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -104,19 +105,15 @@ var _ = Describe("performance", func() {
+				for param, expected := range sysctlMap {
+					By(fmt.Sprintf("executing the command \"sysctl -n %s\" inside the pod %s", param, mcdName))
+					Eventually(func() string {
+						out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
```
looks like the new format of the command is now failing and getting an error upon execution

dorzheh commented:
No,the parameters doesn't exist hence the sysctl command fails

On Sun, Jan 5, 2020, 5:01 PM yanir quinn <notifications@github.com> wrote:

> *@yanirq* commented on this pull request.
> ------------------------------
>
> In features/functests/performance/performance.go
> <https://github.com/openshift-kni/baremetal-deploy/pull/118#discussion_r363098514>
> :
>
> > -				out, err := exec.Command("oc", "rsh", "-n", "openshift-machine-config-operator", mcdName,
> -					"sysctl", "-A").CombinedOutput()
> -				Expect(err).ToNot(HaveOccurred())
> -
> -				for _, str := range strings.Split(string(out), "\n") {
> -					line := strings.Split((string(str)), "=")
> -					param := strings.TrimSpace(line[0])
> -					value := strings.TrimSpace(strings.Join(line[1:], ""))
> -					if expected, ok := sysctlMap[param]; ok {
> -						By(fmt.Sprintf("checking whether parameter %s value is %s", param, expected))
> -						Expect(value).To(Equal(expected), fmt.Sprintf("parameter %s value is not %s", param, expected))
> -					}
> +				for param, expected := range sysctlMap {
> +					By(fmt.Sprintf("executing the command \"sysctl -n %s\" inside the pod %s", param, mcdName))
> +					Eventually(func() string {
> +						out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
>
> looks like the new format of the command is now failing and getting an
> error upon execution
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift-kni/baremetal-deploy/pull/118?email_source=notifications&email_token=ABWDK23NECA5PQWLLSBK5O3Q4HY2VA5CNFSM4J662NKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQV2QLQ#pullrequestreview-338405422>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWDK22XI65ETE4ENBUVOKLQ4HY2VANCNFSM4J662NKA>
> .
>

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +63,145 @@ var _ = Describe("performance", func() {
+					line := strings.TrimSpace(string(out))
+					return strings.Contains(line, perfSysctlKernelVersionValuePreemptEntry) &&
+						strings.Contains(line, perfSysctlKernelVersionValueRtEntry)
+				}, perfSysctlTimeout*time.Second, perfSysctlPollInterval*time.Second, ).Should(BeTrue(), "RT kernel is not installed")
```
The last "," seems redundant or wrong .
In addition the async calls inside a loop using eventually might cause faulty behaviour here.

dorzheh commented:
@Yanir Quinn <yquinn@redhat.com> , the comma will be removed , though it
doesn't  affect . Regarding async calls. Indeed , it works asynchronously
and  sounds like it could be an issue if the test runs inside a  runtime
environment with limited amount of resources. However the loop is pretty
short and apparently each iteration is blocked until a result will be
received from the function.   In any case, everything works like a charm on
my systems and seems the issue is only when it runs by the CI system.


On Mon, Jan 6, 2020, 10:40 AM yanir quinn <notifications@github.com> wrote:

> *@yanirq* commented on this pull request.
> ------------------------------
>
> In features/functests/performance/performance.go
> <https://github.com/openshift-kni/baremetal-deploy/pull/118#discussion_r363200105>
> :
>
> >  				Expect(err).ToNot(HaveOccurred())
> -				Expect(strings.Contains(string(bootLoaderEntries), "iso_initrd.img")).To(BeTrue())
> +				mcdName := mcd.ObjectMeta.Name
> +				By(fmt.Sprintf("executing the command \"sysctl -n %s\" inside the pod %s", perfSysctlKernelVersionParam, mcdName))
> +				Eventually(func() bool {
> +					out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
> +						"-c", perfMachineConfigDaemonContainer, mcdName, "sysctl", "-n", perfSysctlKernelVersionParam).CombinedOutput()
> +					Expect(err).ToNot(HaveOccurred())
> +					line := strings.TrimSpace(string(out))
> +					return strings.Contains(line, perfSysctlKernelVersionValuePreemptEntry) &&
> +						strings.Contains(line, perfSysctlKernelVersionValueRtEntry)
> +				}, perfSysctlTimeout*time.Second, perfSysctlPollInterval*time.Second, ).Should(BeTrue(), "RT kernel is not installed")
>
> The last "," seems redundant or wrong .
> In addition the async calls inside a loop using eventually might cause
> faulty behaviour here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift-kni/baremetal-deploy/pull/118?email_source=notifications&email_token=ABWDK24HKT5YLZ3QNZDFGE3Q4LU5RA5CNFSM4J662NKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQWWVTI#pullrequestreview-338520781>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWDK27SJWFBPMZUGLVTSHDQ4LU5RANCNFSM4J662NKA>
> .
>

yanirq commented:
@dorzheh  Perhaps we should add a check for successful deployment as we do for the machineconfigs in https://github.com/openshift-kni/baremetal-deploy/blob/d7a57a0f2f948d1bdecf9b90db238049c7cdc813/features/performance/deploy.sh#L59 

Not necessarily the same way, maybe we can use Node tuning operator for that.

dorzheh commented:
@Yanir Quinn <yquinn@redhat.com> , do we have an opportunity to execute CI
jobs within a private repo?

On Mon, Jan 6, 2020 at 12:26 PM yanir quinn <notifications@github.com>
wrote:

> @dorzheh <https://github.com/dorzheh> Perhaps we should add a check for
> successful deployment as we do for the machineconfigs in
> https://github.com/openshift-kni/baremetal-deploy/blob/d7a57a0f2f948d1bdecf9b90db238049c7cdc813/features/performance/deploy.sh#L59
>
> Not necessarily the same way, maybe we can use Node tuning operator for
> that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift-kni/baremetal-deploy/pull/118?email_source=notifications&email_token=ABWDK25D37XX4W47DPS4Z6LQ4MBOBA5CNFSM4J662NKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIFBNXA#issuecomment-571086556>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWDK25ZZZGWLRBTBMCZUZTQ4MBOBANCNFSM4J662NKA>
> .
>

yanirq commented:
> @yanir Quinn <yquinn@redhat.com> , do we have an opportunity to execute CI jobs within a private repo?

@fedepaol ?

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +63,145 @@ var _ = Describe("performance", func() {
+
+			for _, node := range nodes.Items {
+				mcd, err := mcdForNode(&node)
+				Expect(err).ToNot(HaveOccurred())
```
can you return false here instead of Expect(err).ToNot ?

slintes replied:
no, not here, a few lines below in the `Eventually` block. There you need to return a string which does not fullfill the expectation, in order to rerun the oc command. `Expect(err).ToNot(HaveOccurred())` does fail the test and not retry.

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +63,145 @@ var _ = Describe("performance", func() {
+					Eventually(func() string {
+						out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
+							"-c", perfMachineConfigDaemonContainer, mcdName, "sysctl", "-n", param).CombinedOutput()
+						Expect(err).ToNot(HaveOccurred())
```
can you return false here instead of Expect(err).ToNot ?

slintes replied:
You need to return a string here which does not fulfill the expected string

dorzheh replied:
@slintes , i think its better to remove this line (Expect..). In this case we will get more profound output, for example:
    parameter net.core.busy_poll value is not 50
    Expected
        <string>: sysctl: cannot stat /proc/sys/net/core/busy_poll: No such file or directory\ncommand terminated with exit code 255
    to equal
        <string>: 50

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +63,145 @@ var _ = Describe("performance", func() {
+				Eventually(func() bool {
+					out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
+						"-c", perfMachineConfigDaemonContainer, mcdName, "sysctl", "-n", perfSysctlKernelVersionParam).CombinedOutput()
+					Expect(err).ToNot(HaveOccurred())
```
can you return false here instead of Expect(err).ToNot ?

slintes commented on features/functests/performance/performance.go:
```diff
@@ -50,20 +63,145 @@ var _ = Describe("performance", func() {
+					Eventually(func() string {
+						out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
+							"-c", perfMachineConfigDaemonContainer, mcdName, "sysctl", "-n", param).CombinedOutput()
+						Expect(err).ToNot(HaveOccurred())
```
here

yanirq commented:
@dorzheh  the way our deployment works (as demonstrated in CI) is after applying all the machine configs and tuning profiles it performs a reboot to the node so some aspects of the real-time profiles might not be updated yet. 
We can either enter a sleep/delay at the beginning of the tests or even better check mcdForNode for running state

yanirq commented:
@dorzheh please rebase the PR on master

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -1,73 +1,214 @@
+		const perfWorkerRtKernelArgsMc = "12-worker-rt-kargs"
+		It("Should contain Machine Configuration profile "+perfWorkerRtKernelArgsMc, func() {
+			_, err := clients.MachineConfig.MachineconfigurationV1().MachineConfigs().Get(perfWorkerRtKernelArgsMc, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("doesn't contain %s Machine Configiguration profile", perfWorkerRtKernelArgsMc))
```
typo : "Configiguration"

dorzheh replied:
@yanirq . Thanks

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -1,73 +1,214 @@
-			for _, node := range nodes.Items {
+	// 11-pre-boot-worker-tuning.yaml related verification
+	var _ = Context("Pre boot tuning adjusted by the Machine Config Operator ", func() {
+		const perfWorkerPreBootTuning = "11-worker-pre-boot-tuning"
```
where did "11-worker-pre-boot-tuning" came from ?  did you mean "11-worker-rt-kernel" ?

yanirq replied:
notice that both rt kernel and pre boot scripts are set on the same machine config

dorzheh replied:
@yanirq , ah , forgot to rebase my stuff.

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -1,73 +1,214 @@
+				By("executing the command \"uname -v\" inside the pod " + mcdName)
+				Eventually(func() bool {
+					out, err := exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace,
+						"-c", perfMachineConfigDaemonContainer, mcdName, "uname", "-v").CombinedOutput()
```
is there a chance we could miss the PREEMPT and RT values for using -v instead of -a ?
(for uname,  -a will cover everything)

dorzheh replied:
@yanirq , no uname -v works fine in our case

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -1,73 +1,214 @@
+
+	// 12-worker-rt-kargs.yaml related verifications
+	var _ = Context("RT kernel arguments", func() {
+		const perfWorkerRtKernelArgsMc = "12-worker-rt-kargs"
```
this was changes to 10-worker-rt-kargs

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -1,73 +1,214 @@
+				"kernel.sched_migration_cost_ns":  "5000000",
+			}
+			const perfNodeNetworkLatencyTuned = "openshift-node-network-latency"
+			_, err := clients.CntoConfig.TunedV1().Tuneds(perfClusterNodeTuningOperatorNamespaces).Get(perfNodeNetworkLatencyTuned,metav1.GetOptions{})
```
im not a big fan of adding a lot of dependencies for this one liner (if this is the case) if the check turns out to be un-needed then lets remove it.

dorzheh replied:
@yanirq , perhaps. The other way  - execute the oc command and parse the output. Lets the CI to complete all the tests successfully and afterwards decide about more convenient solution.

yanirq commented on features/functests/performance/performance.go:
```diff
@@ -254,12 +229,12 @@ func checkWorkerRtProfileReadiness(nodes []k8sv1.Node) {
+		By(fmt.Sprintf("Searching for the file %s.Executing the command \"ls %s\" inside the pod %s", file, file, mcdName))
 		err = exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace, "-c",
 			perfMachineConfigDaemonContainer, mcdName, "ls", "/rootfs/"+file).Run()
 		Expect(err).To(BeNil(), "cannot find the script "+file)
```
why are we checking err=nil here ?

yanirq replied:
why are we checking err=nil here ?

gsr-shanks commented on features/functests/performance/performance.go:
```diff
@@ -77,17 +218,51 @@ func mcdForNode(node *k8sv1.Node) (*k8sv1.Pod, error) {
+		By(fmt.Sprintf("executing the command \"ls %s\" inside the pod %s", file, mcdName))
+		err = exec.Command("oc", "rsh", "-n", mcd.ObjectMeta.Namespace, "-c",
+			perfMachineConfigDaemonContainer, mcdName, "ls", "/rootfs/"+file).Run()
+		Expect(err).To(BeNil(), "cannot find the script "+file)
```
Shouldn't the err read as "cannot find the file " instead of "cannot find the script "?

yanirq commented:
/retest

gsr-shanks commented on install-steps.md:
```diff
@@ -603,17 +609,17 @@ to fix the ConfigMap for the Metal3 operator. This ConfigMap is used to notify
       http_port: "6180"
-      ironic_endpoint: http://172.22.0.3:6385/v1/
+      ironic_endpoint: http://<provisioning_ip>:6385/v1/
       ironic_inspector_endpoint: http://172.22.0.3:5050/v1/
```
Maybe we would want to remove the above IP as well.

yanirq replied:
@gsr-shanks @dorzheh this looks like it doesn't belong to this PR. Another rebase against the master branch should align the files.

yanirq commented:
/retest

yanirq commented:
/retest

dorzheh commented:
/retest

yanirq commented:
/lgtm
/approve

yanirq commented:
/approve

dorzheh commented:
Closing this PR. Suppressed by #124
